File: backend/routes/modules.py
import os, json
from flask import Blueprint, jsonify, abort
from models.auth import login_required

modules_bp = Blueprint('modules', __name__)
CONTENT_DIR = os.path.join(os.path.dirname(__file__), '..', 'content', 'modules')
MODULE_ORDER = ['sql-injection','xss','insecure-auth','path-traversal','buffer-overflow']

def load_module(mid):
    path = os.path.join(CONTENT_DIR, mid.replace('-','_') + '.json')
    if not os.path.exists(path):
        return None
    with open(path, encoding='utf-8') as f:
        return json.load(f)

@modules_bp.route('/', methods=['GET'])
@login_required
def get_all_modules():
    modules = []
    for mid in MODULE_ORDER:
        m = load_module(mid)
        if m:
            modules.append({
                'id': m['id'], 'title': m['title'],
                'owasp_category': m['owasp_category'], 'cwe': m.get('cwe',[]),
                'difficulty': m['difficulty'], 'estimated_time': m['estimated_time'],
                'description': m['description'],
                'objective_count': len(m.get('learning_objectives',[])),
                'section_count': len(m.get('sections',[])),
                'quiz_count': len(m.get('quiz',[])),
            })
    return jsonify({'modules': modules, 'total': len(modules)})

@modules_bp.route('/<module_id>', methods=['GET'])
@login_required
def get_module(module_id):
    m = load_module(module_id)
    if not m:
        abort(404)
    return jsonify(m)

File: frontend/src/pages/Modules.js
import React, { useEffect, useState } from 'react';
import { useNavigate } from 'react-router-dom';
import { getModules } from '../services/api';
import './Modules.css';

const DIFFICULTY_COLOUR = {
  Beginner:     'badge-success',
  Intermediate: 'badge-warning',
  Advanced:     'badge-danger',
};

const OWASP_ICONS = {
  'sql-injection':  '💉',
  'xss':            '📜',
  'insecure-auth':  '🔐',
  'path-traversal': '📂',
  'buffer-overflow':'💾',
};

export default function Modules() {
  const [modules, setModules]   = useState([]);
  const [loading, setLoading]   = useState(true);
  const [error, setError]       = useState(null);
  const [filter, setFilter]     = useState('All');
  const navigate = useNavigate();

  useEffect(() => {
    getModules()
      .then(res => setModules(res.data.modules))
      .catch(() => setError('Could not load modules. Is the backend running?'))
      .finally(() => setLoading(false));
  }, []);

  const difficulties = ['All', 'Beginner', 'Intermediate', 'Advanced'];
  const visible = filter === 'All'
    ? modules
    : modules.filter(m => m.difficulty === filter);

  if (loading) return <div className="page"><div className="loader">Loading modules…</div></div>;
  if (error)   return <div className="page"><div className="error-box">{error}</div></div>;

  return (
    <div className="page modules-page">
      <div className="modules-header">
        <div>
          <h1>Learning Modules</h1>
          <p className="text-muted">
            Five OWASP Top 10 vulnerability types — each with explanations, code examples, and a quiz.
          </p>
        </div>
        <div className="filter-bar">
          {difficulties.map(d => (
            <button
              key={d}
              className={`filter-btn ${filter === d ? 'active' : ''}`}
              onClick={() => setFilter(d)}
            >
              {d}
            </button>
          ))}
        </div>
      </div>

      <div className="modules-grid">
        {visible.map((mod, i) => (
          <div
            key={mod.id}
            className="module-card"
            onClick={() => navigate(`/modules/${mod.id}`)}
            style={{ animationDelay: `${i * 0.07}s` }}
          >
            <div className="module-card-icon">{OWASP_ICONS[mod.id] || '🛡️'}</div>
            <div className="module-card-body">
              <div className="module-card-meta">
                <span className={`badge ${DIFFICULTY_COLOUR[mod.difficulty]}`}>
                  {mod.difficulty}
                </span>
                <span className="text-muted module-time">⏱ {mod.estimated_time}</span>
              </div>
              <h3 className="module-card-title">{mod.title}</h3>
              <p className="module-card-owasp">{mod.owasp_category}</p>
              <p className="module-card-desc">{mod.description}</p>
              <div className="module-card-stats">
                <span>📚 {mod.section_count} sections</span>
                <span>❓ {mod.quiz_count} quiz questions</span>
              </div>
              <div className="module-card-cwe">
                {mod.cwe?.map(c => (
                  <span key={c} className="cwe-tag">{c}</span>
                ))}
              </div>
            </div>
            <div className="module-card-arrow">→</div>
          </div>
        ))}
      </div>

      {visible.length === 0 && (
        <div className="empty-state">No modules match this filter.</div>
      )}
    </div>
  );
}
